crates/models/mpt/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+//! An implementation of [GPT-2](https://huggingface.co/docs/transformers/model_doc/gpt2) for the `llm` ecosystem.
+#![deny(missing_docs)]
+
+use ggml::Tensor;
+use llm_base::{
+    ggml,
+    model::{common, HyperparametersWriteError},
+    util, FileType, InferenceParameters, InferenceSession, InferenceSessionConfig, KnownModel,
+    LoadError, ModelParameters, OutputRequest, TokenId, Vocabulary,
+};
+
+/// The  MosaicPretrainedTransformer (MPT) model. Ref: [Mosaic ML](https://huggingface.co/mosaicml)
+///
+/// # Safety
+/// This implements [Send] and [Sync] as it is immutable after construction.
+pub struct Mpt {
+    hyperparameters: Hyperparameters,
+    n_context_tokens: usize,
+    vocabulary: Vocabulary,
+    ln_f_g: Tensor,
+    ln_f_b: Tensor,
+    wte: Tensor,
+    wpe: Tensor,
+    lm_head: Tensor,
+    layers: Vec<Layer>,
+    inference_params: InferenceParameters,
+    _context: ggml::Context,
+    _mmap: Option<llm_base::Mmap>,
+}
+unsafe impl Send for Mpt {}
+unsafe impl Sync for Mpt {}
+
+impl KnownModel for Mpt {
+    type Hyperparameters = Hyperparameters;
+
+    fn new<E: std::error::Error>(
+        hyperparameters: Self::Hyperparameters,
+        params: ModelParameters,
+        vocabulary: Vocabulary,
+        tensor_loader: impl llm_base::TensorLoader<E>,
+    ) -> Result<Self, E> {
+       todo!("implement this")
+    }
+
+    fn start_session(&self, config: InferenceSessionConfig) -> InferenceSession {
+        todo!("implement this")
+    }
+
+    fn evaluate(
+        &self,
+        session: &mut InferenceSession,
+        params: &InferenceParameters,
+        input_tokens: &[TokenId],
+        output_request: &mut OutputRequest,
+    ) {
+        todo!("implement this")
+    }
+
+    /// Returns the vocabulary used by this model.
+    fn vocabulary(&self) -> &Vocabulary {
+        &self.vocabulary
+    }
+
+    fn n_context_tokens(&self) -> usize {
+        self.n_context_tokens
+    }
+
+    fn bot_token_id(&self) -> Option<TokenId> {
+        todo!("implement this")
+    }
+
+    fn eot_token_id(&self) -> TokenId {
+        todo!("implement this")
+    }
+
+    fn inference_parameters(&self) -> &InferenceParameters {
+        &self.inference_params
+    }
+
+}
+
+/// MPT [hyperparameters](https://en.wikipedia.org/wiki/Hyperparameter_(machine_learning))
+#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
+pub struct Hyperparameters {
+    //for the love of god please rename these to something more descriptive and standardise the names accross the different models 
```

Comment:
Yeah, the names leave a lot to be desired, but for now I have taken the approach of trying to mimic the nomenclature used in the "reference" C++ implementation in order to help make debugging easier for myself. Another thing is that in many (most) cases I don't actually know the purpose of the data...I've tried to add documentation to the ones that I feel comfortable doing so for.

Reply:
Yeah, we should definitely go through and fix these up. I think a lot of the original names can be recovered from looking at the Python / HF model docs.